File: ReactAndroid/src/main/java/com/facebook/hermes/instrumentation/HermesSamplingProfiler.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.hermes.instrumentation;

import com.facebook.soloader.SoLoader;

/** Hermes sampling profiler static JSI API. */
public class HermesSamplingProfiler {
  static {
    SoLoader.loadLibrary("jsijniprofiler");
  }

  /** Start sample profiling. */
  public static native void enable();

  /** Stop sample profiling. */
  public static native void disable();

  /**
   * Dump sampled stack traces to file.
   *
   * @param filename the file to dump sampling trace to.
   */
  public static native void dumpSampledTraceToFile(String filename);

  private HermesSamplingProfiler() {}
}

File: ReactAndroid/src/main/java/com/facebook/hermes/instrumentation/HermesSamplingProfiler.cpp
// Copyright 2004-present Facebook. All Rights Reserved.

#include "HermesSamplingProfiler.h"

#include <hermes/hermes.h>

namespace facebook {
namespace jsi {
namespace jni {

void HermesSamplingProfiler::enable(jni::alias_ref<jclass>) {
  hermes::HermesRuntime::enableSamplingProfiler();
}

void HermesSamplingProfiler::disable(jni::alias_ref<jclass>) {
  hermes::HermesRuntime::disableSamplingProfiler();
}

void HermesSamplingProfiler::dumpSampledTraceToFile(
    jni::alias_ref<jclass>,
    std::string filename) {
  hermes::HermesRuntime::dumpSampledTraceToFile(filename);
}

void HermesSamplingProfiler::registerNatives() {
  javaClassLocal()->registerNatives({
      makeNativeMethod("enable", HermesSamplingProfiler::enable),
      makeNativeMethod("disable", HermesSamplingProfiler::enable),
      makeNativeMethod(
          "dumpSampledTraceToFile",
          HermesSamplingProfiler::dumpSampledTraceToFile),
  });
}

} // namespace jni
} // namespace jsi
} // namespace facebook
